webhook/resources/node/validator.go

```diff
@@ -0,0 +1,149 @@
+package node
+
+import (
+	"fmt"
+	"math"
+
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/util"
+	"github.com/longhorn/longhorn-manager/webhook/admission"
+	werror "github.com/longhorn/longhorn-manager/webhook/error"
+	"github.com/sirupsen/logrus"
+)
+
+type nodeValidator struct {
+	admission.DefaultValidator
+	ds *datastore.DataStore
+}
+
+func NewValidator(ds *datastore.DataStore) admission.Validator {
+	return &nodeValidator{ds: ds}
+}
+
+func (n *nodeValidator) Resource() admission.Resource {
+	return admission.Resource{
+		Name:       "nodes",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   longhorn.SchemeGroupVersion.Group,
+		APIVersion: longhorn.SchemeGroupVersion.Version,
+		ObjectType: &longhorn.Node{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Update,
+			admissionregv1.Delete,
+		},
+	}
+}
+
+func (n *nodeValidator) Update(request *admission.Request, oldObj runtime.Object, newObj runtime.Object) error {
+	oldNode := oldObj.(*longhorn.Node)
+	newNode := newObj.(*longhorn.Node)
+
+	// Only scheduling disabled node can be evicted
+	// Can not enable scheduling on an evicting node
+	if oldNode.Spec.EvictionRequested && oldNode.Spec.AllowScheduling {
+		return werror.NewInvalidError(fmt.Sprintf("need to disable scheduling on node %v for node eviction, or cancel eviction to enable scheduling on this node",
+			oldNode.Name), "")
+	}
+
+	// We need to make sure the tags passed in are valid before updating the node.
+	_, err := util.ValidateTags(newNode.Spec.Tags)
+	if err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if newNode.Spec.EngineManagerCPURequest < 0 || newNode.Spec.ReplicaManagerCPURequest < 0 {
+		return werror.NewInvalidError(fmt.Sprintf("found invalid EngineManagerCPURequest %v or ReplicaManagerCPURequest %v during node %v update",
+			newNode.Spec.EngineManagerCPURequest, newNode.Spec.ReplicaManagerCPURequest, oldNode.Name), "")
+	}
+
+	if newNode.Spec.EngineManagerCPURequest != 0 || newNode.Spec.ReplicaManagerCPURequest != 0 {
+		kubeNode, err := n.ds.GetKubernetesNode(oldNode.Name)
+		if err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+		allocatableCPU := float64(kubeNode.Status.Allocatable.Cpu().MilliValue())
+		engineManagerCPUSetting, err := n.ds.GetSetting(types.SettingNameGuaranteedEngineManagerCPU)
+		if err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+		engineManagerCPUInPercentage := engineManagerCPUSetting.Value
+		if newNode.Spec.EngineManagerCPURequest > 0 {
```

Comment:
`newNode.Spec.EngineManagerCPURequest` should be always > 0 here, so probably we don't need to have if check here anymore.

Reply:
https://github.com/longhorn/longhorn-manager/blob/master/k8s/pkg/apis/longhorn/v1beta2/node.go#L79

Should replace `Minimum=0` with `exclusiveMinimum=0`

Reply:
https://github.com/longhorn/longhorn-manager/blob/master/upgrade/v110to111/upgrade.go#L115
Also need to mutate the `EngineManagerCPURequest` with zero value

Reply:
https://github.com/longhorn/longhorn-manager/blob/555b5a38f57b1875af37614467623010f4d799b8/controller/controller_manager.go

If `EngineManagerCPURequest` or `ReplicaManagerCPURequest` is 0, the default value will be set.
Thus, `0` is also available value here.